### PR TITLE
[in_app_purchase_storekit] Add support for purchase and transactions #7574

### PR DESCRIPTION
**Pull request by LouiseHsu**
This PR contains the preliminary implementations for handling [purchases](https://developer.apple.com/documentation/storekit/product/3791971-purchase) and [transactions](https://developer.apple.com/documentation/storekit/transaction) 

Of note, as of StoreKit 2, the [result of a call to a purchase will only return a Transaction](https://developer.apple.com/documentation/storekit/product/purchaseresult) when the purchase is successful. If the status of purchase is pending, or cancelled, nothing will be returned. 

This is in contrast to StoreKit 1, where developers could check the state of a Transaction with [TransactionState](https://developer.apple.com/documentation/storekit/skpaymenttransaction/1411275-transactionstate). This means that migrating to StoreKit 2 will require developers to update their handling of transactions accordingly. 

Relands https://github.com/flutter/packages/pull/7574

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

### CODE REVIEW

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.swift [hunk 1]
```diff
@@ -27,6 +27,21 @@ public class InAppPurchasePlugin: NSObject, FlutterPlugin, InAppPurchaseAPI {
   // This property is optional, as it requires self to exist to be initialized.
   public var paymentQueueHandler: FLTPaymentQueueHandlerProtocol?
 
+  // This should be an Task, but Task is on available >= iOS 13
+  private var _updateListenerTask: Any?
+
+  @available(iOS 13.0, *)
+  var getListenerTaskAsTask: Task<(), Never> {
```
Comment (hellohuanlin) [inline packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.swift:34]: this should be called `var updateListenerTask: Task<(), Never>` (no `get` in getters naming convention)

**Comment by stuartmorgan** [comment]
> Relands #7574

For future reference, what were the changes to resolve the issue that led to the revert?

**Comment by stuartmorgan** [comment]
This failed `stable` post-submit again: https://github.com/flutter/packages/runs/31667628631

**Comment by stuartmorgan** [comment]
> Even thought storekittest is technically a mac 11 framework, the bulk of the testing features, such simulating network errors and things like that are mac 14 only. The presubmits passed because it ran on a Mac 14 bot, but the post submit failed because it ran on a Mac 13 bot. :(

I *think* that if in your next version of this you temporarily, in your PR, change the `.ci.yaml` to only allow Mac 13 hosts, you might be able to get presubmit testing to make sure everything is good. (I'm not 100% sure that change takes effect in presubmit though.)

**Comment by LouiseHsu** [comment]
> > Even thought storekittest is technically a mac 11 framework, the bulk of the testing features, such simulating network errors and things like that are mac 14 only. The presubmits passed because it ran on a Mac 14 bot, but the post submit failed because it ran on a Mac 13 bot. :(
> 
> I _think_ that if in your next version of this you temporarily, in your PR, change the `.ci.yaml` to only allow Mac 13 hosts, you might be able to get presubmit testing to make sure everything is good. (I'm not 100% sure that change takes effect in presubmit though.)

Oh, thats a good idea. Im really struggling with testing this properly cuz I have to keep restarting and trying to get specific bots.

**Pull request merged**